code/binary-tree/ViewModel/src/main/java/ru/unn/agile/binarytree/viewmodel/ViewModel.java

```diff
@@ -0,0 +1,217 @@
+
+    public ViewModel() {
+        addKey.set("");
+        addValue.set("");
+        addStatus.set(Status.WAITING.toString());
+
+        findKey.set("");
+        findResult.set("");
+        findStatus.set(Status.WAITING.toString());
+
+        removeKey.set("");
+        removeResult.set("");
+        removeStatus.set(Status.WAITING.toString());
+
+        final List<StringProperty> fields = new ArrayList<>() {
+            {
+                add(addKey);
+                add(addValue);
+                add(findKey);
+                add(removeKey);
+            }
+        };
+
+        for (StringProperty field : fields) {
+            final ValueChangeListener listener = new ValueChangeListener();
+            field.addListener(listener);
+            valueChangedListeners.add(listener);
+        }
+    }
+
+    public StringProperty addKeyProperty() {
+        return addKey;
+    }
+
+    public StringProperty addValueProperty() {
+        return addValue;
+    }
+
+    public StringProperty addStatusProperty() {
+        return addStatus;
+    }
+
+    public StringProperty findKeyProperty() {
+        return findKey;
+    }
+
+    public StringProperty findResultProperty() {
+        return findResult;
+    }
+
+    public StringProperty findStatusProperty() {
+        return findStatus;
+    }
+
+    public StringProperty removeKeyProperty() {
+        return removeKey;
+    }
+
+    public StringProperty removeResultProperty() {
+        return removeResult;
+    }
+
+    public StringProperty removeStatusProperty() {
+        return removeStatus;
+    }
+
+    public String getAddStatus() {
+        return addStatus.get();
+    }
+
+    public String getFindResult() {
+        return findResult.get();
+    }
+
+    public String getFindStatus() {
+        return findStatus.get();
+    }
+
+    public String getRemoveResult() {
+        return removeResult.get();
+    }
+
+    public String getRemoveStatus() {
+        return removeStatus.get();
+    }
+
+    public void add() {
+        final int key = Integer.parseInt(addKey.get());
+        final String value = addValue.get();
+
+        storage.add(key, value);
+
+        addStatus.set(Status.SUCCESS.toString());
+    }
+
+    public void find() {
+        final int key = Integer.parseInt(findKey.get());
+
+        final String result = storage.find(key);
+
+        findResult.set(result);
+        findStatus.set(Status.SUCCESS.toString());
+    }
+
+    public void remove() {
+        final int key = Integer.parseInt(removeKey.get());
+
+        final boolean result = storage.remove(key);
+
+        removeResult.set(Boolean.toString(result));
+        removeStatus.set(Status.SUCCESS.toString());
+    }
+
+    private Status getAddInputStatus() {
+        Status addStatus = Status.READY;
+        if (addKey.get().isEmpty() || addValue.get().isEmpty()) {
+            addStatus = Status.WAITING;
+        }
```

Comment:
Even if status is `WAITING` we still can add only key, because in `add` function doesn have check that `value` filed is not null. Maybe we should check status before adding?

Reply:
Ideally, we should add disability feature to the form (but it's a matter of the extra task, as described in the task description by Kirill) in case it's WAITING.

Because I don't want to get extra knowledge about this in binding - I propose a solution which fits in the best way for the potential disability (WAITING if one of fields is not set)

By the way, as I see from other PR and code snippets, we are not going to check the statuses in the `add/find/remove` operations. Simply, we should assume that add, find and remove operations has a precondition that forms must be filled.

It's the simplest form of contract with user, but it has a right to exist. In other words, we can put a word in documentation about this :)

Reply:
ok